walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,42 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.sql.Time;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    SortedSet<TimeRange> timeRangesSorted = new TreeSet<TimeRange>(TimeRange.ORDER_BY_START);
+    for(Event e: events){ //some map could be used here
+      if(hasAttendee(e, request.getAttendees())){
+        timeRangesSorted.add(e.getWhen());
+      }
+    }
+
+    TimeRange guardian = TimeRange.fromStartDuration(TimeRange.END_OF_DAY+1, 0); //Shouldn't timerange set outside the minutes of the day be invalid?
+    timeRangesSorted.add(guardian); //Could be done by an additional if at the end as well
+
+    int lastFinish = TimeRange.START_OF_DAY;
+    Collection<TimeRange> appropriateTimeranges = new ArrayList<TimeRange>();
+    for(TimeRange nextEvent: timeRangesSorted){
+      if(lastFinish < nextEvent.start()){ //should I extract nextEvent.start() to a local variable?
+        TimeRange empty = TimeRange.fromStartEnd(lastFinish, nextEvent.start(), false);
+        if(empty.duration() >= request.getDuration()){
+          appropriateTimeranges.add(empty);
+        }
+      }
+      if(lastFinish < nextEvent.end()){
+        lastFinish = nextEvent.end();
+      }
+    }
+
+    return appropriateTimeranges;
+  }
+
+  public boolean hasAttendee(Event event, Collection<String> Attendees){//will it delete the attendees if the event?
+    Collection<String> eventAttendees = new HashSet<String>();
+    eventAttendees.addAll(event.getAttendees());
+    eventAttendees.retainAll(Attendees);
+    return eventAttendees.size() > 0;
```

Comment:
This function seems correct, but could be optimized by using e.g. https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection)

Reply:
Just a question out of curiousity: Are such general functions (that can handle any collection) optimized for specific data structures? (For example in this case looking up if one element is in the set is easy, but if the set was a list, you had to iterate over it.)

Reply:
Such functions generally use other collection methods, so their performance depends on the type of parameters and exact inner algorithm. But generally they do try to be as fast as possible (e.g. disjoint check if any parameter is a set or other constant-time-check collection and iterates over the other one).

You're completely right with your conclusion - since we have a set here, it will be effective in checking if a set contains an element. With lists it would be O(n2), because of iterations.

Reply:
Actually in this case they can achieve high performance with any collection, if they first create a HashSet from both collections (just one iteration), and then just do the same as with HashSets